src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -50,6 +51,176 @@ static void Main()
                 );
         }
 
+        [Fact]
+        public void SpeakableInference_MethodTypeInference()
+        {
+            var source =
+@"class Program
+{
+    void M<T>(T t)
+    {
+        if (t == null) throw null;
+        t.ToString();
+        var t2 = Copy(t);
+        t2.ToString(); // warn
+    }
+    static T Copy<T>(T t) => throw null;
+}";
+            var comp = CreateCompilationWithIndexAndRange(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         t2.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t2").WithLocation(8, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("void M(int? t)")]
+        [InlineData("void M<T>(T? t) where T : struct")]
+        public void SpeakableInference_MethodTypeInference_NullableValueType(string signature)
+        {
+            var source =
+@"class Program
+{
+    SIGNATURE
```

Comment:
>SIGNATURE [](start = 4, length = 9)

Consider writing two tests (with no text substitution), or two methods in this same test.